x/consensus/keeper/estimate.go

```diff
@@ -0,0 +1,162 @@
+		if err != nil {
+			logger.WithError(err).Warn("Failed to get supported queues")
+			continue
+		}
+		for _, opt := range opts {
+			msgs, err := k.GetMessagesFromQueue(sdkCtx, opt.QueueTypeName, 0)
+			if err != nil {
+				logger.WithError(err).Warn("Failed to get messages from queue")
+				continue
+			}
+
+			logger = logger.WithChain(opt.ChainReferenceID)
+			cq, err := k.getConsensusQueue(sdkCtx, opt.QueueTypeName)
+			if err != nil {
+				logger.WithError(err).Warn("Failed to get consensus queue")
+				continue
+			}
+			for _, msg := range msgs {
+				logger = logger.WithFields("msg_id", msg.GetId())
+				if err := k.checkAndProcessEstimatedMessage(ctx, msg, cq); err != nil {
+					logger.WithError(err).Warn("Failed to process estimated message")
+					continue
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (k Keeper) checkAndProcessEstimatedMessage(ctx context.Context,
+	msg types.QueuedSignedMessageI,
+	q consensus.Queuer,
+) error {
+	_, rcid := q.ChainInfo()
+	logger := liblog.FromKeeper(ctx, k).WithComponent("check-and-process-estimated-messages").WithChain(rcid)
+
+	// Skip messages that don't require gas estimation
+	if !msg.GetRequireGasEstimation() {
+		return nil
+	}
+
+	// Skip messages that don't have gas estimates
+	if len(msg.GetGasEstimates()) < 1 {
+		return nil
+	}
+	logger = logger.WithFields(
+		"id", msg.GetId(),
+		"nonce", msg.Nonce())
+	logger.Debug("Processing gas estimates for message.")
+
+	estimate, err := k.consensusChecker.VerifyGasEstimates(ctx, k,
+		slice.Map(msg.GetGasEstimates(), func(ge *types.GasEstimate) libcons.GasEstimate {
+			return ge
+		}))
+	if err != nil {
+		if errors.Is(err, libcons.ErrConsensusNotAchieved) {
+			logger.WithError(err).Debug("Consensus not achieved for gas estimates")
+			return nil
+		}
+		return fmt.Errorf("failed to verify gas estimates: %w", err)
+	}
+
+	if err := q.SetElectedGasEstimate(ctx, msg.GetId(), estimate); err != nil {
+		return fmt.Errorf("failed to set elected gas estimate: %w", err)
+	}
+
+	if err := k.checkAndProcessEstimatedSubmitLogicCall(ctx, msg, q, estimate); err != nil {
+		return fmt.Errorf("failed to process estimated submit logic call: %w", err)
+	}
+
+	return nil
+}
+
+func (k Keeper) checkAndProcessEstimatedSubmitLogicCall(
+	ctx context.Context,
+	msg types.QueuedSignedMessageI,
+	q consensus.Queuer,
+	estimate uint64,
+) error {
+	m, err := libmsg.ToEvmMessage(msg, k.cdc)
+	if err != nil {
+		return fmt.Errorf("failed to convert message to evm message: %w", err)
+	}
+	action, ok := m.Action.(*evmtypes.Message_SubmitLogicCall)
+	if !ok {
+		// Skip messages that are not SubmitLogicCall
+		return nil
+	}
+
+	fees, err := k.calculateFeesForEstimate(ctx, sdk.ValAddress(m.GetAssignee()), m.GetChainReferenceID(), estimate)
+	if err != nil {
+		return fmt.Errorf("failed to calculate fees for estimate: %w", err)
+	}
+	action.SubmitLogicCall.Fees = fees
+	_, err = q.Put(ctx, msg, &consensus.PutOptions{
+		MsgIDToReplace: msg.GetId(),
```

Comment:
Do we need to use the same ID here? Could we drop this message and create a new one instead? Perhaps keeping the original ID in a different field.

My worry is that by using the same ID we could get some slow node to try to add it's signature or estimate to a pre-estimated message. Is this even a possibility?

Reply:
This is fully possible, and the current architecture sort of anticipates this. I think recreating the message will be more painful though. In case a new estimate comes in, it will just be disregarded. Same for a new signature against the old state. But this is already the case today, so it's not like we lose something.